### DISCUSSION TITLE
How to disable logger?

### DISCUSSION THREAD

**Discussion by thestumbler** [modm-io/modm#1068] [Q&A]
I tried modifying the board.hpp and startup.cpp files to use the define MODM_BOARD_HAS_LOGGER to conditionally use/exclude things LOGGER. This doesn't seem to work, and the resulting board files generated by LBUILD still suck in logger-related stuff. 

I'm seeing some issues with my code, which primarily access the UART using direct UART read/write calls, and printf. I assume that you shouldn't mix and match, that is have the logger running and also use the UART calls.

EDIT

Okay, I figured out my immediate issue, but it might still be nice to know how to disable logger from the .xml file.

I'm having an assertion on startup for some reason. I was seeing just one letter on the serial port, over and over again. I figured out it was from logging the text message "Assertion...", but the processor reboots before the message is complete. I confirmed this by changing "A" to "O", and then the single letter "O" shows up on the terminal over and over again. Now to track down the culprit causing the assertion. Time to get the debugger up and running.

**Comment by salkinium** [comment]
The `MODM_BOARD_HAS_LOGGER` is just there to tell the application if the BSP defines a logger, it does not control the log level. For that you need to modify [the `MODM_LOG_LEVEL` define](https://docs.modm.io/2023q2/api/stm32f469zit6/group__modm__debug.html#gacdef3d1a3caa5702d1bdf59083b4041e):

```c
#undef  MODM_LOG_LEVEL
#define MODM_LOG_LEVEL modm::log::INFO
```

The debugger should be callable in TUI mode with `scons debug` or `make debug`.

**Reply by salkinium** [reply]
Why are you modifying the `startup.c` file? It shouldn't be necessary, even with a custom linkerscript. The [startup process and customizable hooks are described in detail here](https://modm.io/reference/module/modm-platform-cortex-m/).

**Reply by salkinium** [reply]
The abandonment handler [should actually spin for 2 seconds](https://github.com/modm-io/modm/blob/develop/src/modm/board/board.cpp.in#L47-L54) to give the UART time to print its message. Perhaps that's not enough time. Or something else happens like a hardfault.

**Reply by salkinium** [reply]
Are you [redirecting the printf output correctly](https://modm.io/reference/module/modm-printf/#redirecting-the-output)? We should probably define a weak `putchar_` function in `board.cpp` that redirect to the `Board::Logger`.

**Reply by thestumbler** [reply]
> Why are you modifying the `startup.c` file? It shouldn't be necessary, even with a custom linkerscript.

Just my clumsy attempt to disable logging. I saw that logger was started up there in startup.cpp apparently without any guards. I'll try the correct method you suggest.

**Reply by thestumbler** [reply]
> Are you [redirecting the printf output correctly](https://modm.io/reference/module/modm-printf/#redirecting-the-output)?

I'm pretty sure that I am. I do have a working hello world program from this board's blinky example ( Nucleo STM32H723ZG evaluation board ). In that working code I'm using printf perfectly fine, using the example putchar_ redirect function. 

Is it okay to do both printf and direct UART reads and writes while also having the logger ( assuming your single-threaded application was aware of this so they didn't interfere with each other )?

**Reply by salkinium** [reply]
Yeah, should be fine.

**Comment by thestumbler** [comment]
Quick update, I solved my freezing up issue. This is a boot loader, and I hadn't caught all the various fixed addresses where you read to get certain MCU details. So there were a couple of reads from fixed addresses in memory that didn't exist or were illegal on the H7 which caused a hard fault. At least that's my assumption, since fixing all these locations solved the problem.

Belatedly, I see there is a platform / id / getUniqueId() function that's available, but I don't immediately see the other identifying registers that I was reading (flash size, silicon revision and ID). Now starting to look at actual flash operations, but looks like the H7 isn't included in the list of supported chips for flash programming (issue raised).

**Reply by salkinium** [reply]
Ah yes, the H7 has a completely different MMIO mapping and memory architecture than F7 for example. We have a [pretty cool hardfault catcher module](https://modm.io/reference/module/modm-platform-fault/) btw.
